src/driver/amdxdna/aie2_ctx.c

```diff
@@ -1071,27 +1049,28 @@ static int aie2_add_job_dependency(struct amdxdna_sched_job *job, u32 *syncobj_h
 static void aie2_hwctx_push_job(struct amdxdna_sched_job *job, u64 *seq)
 {
 	struct amdxdna_hwctx *hwctx = job->hwctx;
+	struct amdxdna_sched_job *other;
 	struct dma_fence *fence;
-	long ret;
+	int idx;
 
-again:
 	mutex_lock(&hwctx->priv->io_lock);
-	ret = aie2_hwctx_add_job(hwctx, job);
-	if (ret) {
-		mutex_unlock(&hwctx->priv->io_lock);
-		// Waiting for the first pending cmd to complete before trying again.
-		aie2_cmd_wait(hwctx, hwctx->submitted - HWCTX_MAX_CMDS, 0);
-		goto again;
-	}
+	idx = get_job_idx(hwctx->submitted);
+	other = hwctx->priv->pending[idx];
+	if (other)
+		aie2_job_put(other);
 
 	drm_sched_job_arm(&job->base);
-	fence = dma_fence_get(&job->base.s_fence->finished);
+	hwctx->priv->pending[idx] = job;
+	job->seq = hwctx->submitted++;
 	*seq = job->seq;
+	kref_get(&job->refcnt);
+
+	fence = dma_fence_get(&job->base.s_fence->finished);
+	job->out_fence = fence;
 	drm_sched_entity_push_job(&job->base);
 	mutex_unlock(&hwctx->priv->io_lock);
 
 	aie2_ctx_syncobj_add_fence(hwctx, fence, *seq);
```

Comment:
The fence may have been put and freed already once you pushed the job to drm scheduler before you reach here. You might want to move this line to after line 1070.